### ISSUE TITLE
Cannot find TestStartup

### ISSUE THREAD

**Issue by martincostello** [ivaylokenov/MyTested.AspNetCore.Mvc#269]
I've tried migrating a project from ```1.0.0-preview-final``` to ```1.1.0``` and I just can't get the framework to find the ```TestStartup``` class.

I've tried adding a custom configuration file turning off auto-detection, I've tried specifying the test assembly name explicitly, but nothing seems to work.

It's especially frustrating as the line the exception is coming from is ```MyMvc.StartsFrom<TestStartup>()```.

Any ideas as to how this can be refusing to work? It seems that the static constructor of ```MyMvc``` cannot find the ```testconfig.json``` file or the ```TestStartup``` class even though it's public and in the root namespace of the assembly calling it.

Is it possible that namespaces not matching assembly _names_ is causing the issue?

**Comment by martincostello** [comment]
It's almost as if [this code](https://github.com/ivaylokenov/MyTested.AspNetCore.Mvc/blob/d29e6ccb8b3aa7c9b2bc29e73c161f3f9e021ecf/src/MyTested.AspNetCore.Mvc.Abstractions/Internal/Application/TestApplication.cs#L575-L580) isn't picking the right assembly as the test assembly, and then because it can't find the configuration file (working directory problem?) it throws. Given that this is all in the static constructor, there's no way to successfully call any of the programmatic configuration to get it working.

Making the test assembly initialization lazy would work around any load failures and allow the explicit configuration methods to be usable in this scenario.

**Comment by ivaylokenov** [comment]
@martincostello Hi, can you give me a sample reproduction so that I can investigate? This can be caused if you rename namespaces, projects or directories and there is a mismatch in the default names. 

`Is it possible that namespaces not matching assembly names is causing the issue?` - Yes, it is. Can you check? Because the `TestStartup` type is found by full name in reflection, I need the namespace. No other way to find it besides getting the assembly name. If you have nay suggestion for improving this issue, I will be glad to implement them.

**Comment by martincostello** [comment]
@ivaylokenov Sure. I'll push a branch containing what I've done so far shortly.

**Comment by martincostello** [comment]
@ivaylokenov If you clone the ```[martincostello/api](https://github.com/martincostello/api)``` repo and checkout the ```MyTested-AspNetCore-Mvc-269``` branch and run ```.\Build.ps1 -RestorePackages $true``` you should be able to reproduce.

For reference the assembly name is ```API.Tests``` and the namespace of test startup class is ````MartinCostello.Api`` (assembly names aside, I always put my tests in the same namespaces as the classes they test).

**Comment by ivaylokenov** [comment]
@martincostello 
I have fixed your solution. You can find it attached in this comment. Some remarks to consider:

- If you need different namespaces than the assembly's name, you can use the `testconfig` file to specify the full Startup file location. See the attached solution to understand how.
- I have copied the `hosting.json` and the `appsettings.json` to the test project, because they are required by the `Startup` constructor. You may make them optional or use other mechanics to not need them during the tests.
- Use `MyTested.AspNetCore.Mvc.Universe` in `project.json` if you want to use the full version (with Entity Framework, Caching, Session, etc.). If you want only MVC features, use `MyTested.AspNetCore.Mvc`.
- Instead of `MyMvc.Controller<T>`, you may now use `MyController<T>.Instance()`. Routing has the same custom class.
- Changing the `TestStartup` or the test configuration during async tests is no longer supported because of race conditions that may occur - like here: https://github.com/martincostello/api/blob/master/tests/API.Tests/Controllers/TimeControllerTests.cs#L32. You may either use synchronous tests and change configuration whenever you want (not recommended) or use other methods to provide dependencies for example by using `WithServices`. If you need to reuse a globally registered service, you may call `From.Services<IGlobalService>()`. Check the test in the attached solution to see a working example:

```c#
MyController<TimeController>
                 .Instance()
                 .WithServices(new FakeClock(initial))
```

- You have reached the free quota limits. I have provided a license code for your project in the `testconfig.json` file.
[Fixed-Solution.zip](https://github.com/ivaylokenov/MyTested.AspNetCore.Mvc/files/736873/Fixed-Solution.zip)


If you have any other questions, feel free to ask them! 😄

**Comment by martincostello** [comment]
@ivaylokenov Thank you very much. I did try adding ```testconfig.json``` myself, but I couldn't seem to get it to work. However there were clearly a lot of things that changed since ```1.0.0-preview-final``` that would have taken me an age to get to the bottom of myself and fix. And thank you for the licence! 😃 

I'll close this issue once I've verified everything on my side and committed the changes to my master branch.

**Comment by ivaylokenov** [comment]
@martincostello Sure, take your time. If you find anything else, let me know! A lot has changed since the preview but it's for good I promise. Also, there will be no more breaking changes from now on (except if completely necessary - like coming from the MVC framework itself).

**Comment by martincostello** [comment]
@ivaylokenov Looks like I have a new problem now. It appears that the license validation [doesn't work on Travis CI](https://travis-ci.org/martincostello/api/builds/196196127).

```
System.PlatformNotSupportedException : Operation is not supported on this platform.
Stack Trace:
  at System.Security.Cryptography.RSACryptoServiceProvider..ctor(Int32 dwKeySize)
  at MyTested.AspNetCore.Mvc.Licensing.LicenseValidator.Validate(String license, DateTime releaseDate, String projectNamespace)
  at MyTested.AspNetCore.Mvc.Licensing.LicenseValidator.Validate(IEnumerable`1 licenses, DateTime releaseDate, String projectNamespace)
  at MyTested.AspNetCore.Mvc.Licensing.TestCounter.IncrementAndValidate()
  at MyTested.AspNetCore.Mvc.Builders.Actions.ShouldHave.ShouldHaveTestBuilder`1.SetBuilder()
  at MyTested.AspNetCore.Mvc.Builders.Base.BaseTestBuilderWithComponentBuilder`1.get_Builder()
  /home/travis/build/martincostello/api/tests/API.Tests/Controllers/ToolsControllerTests.cs(42,0): at MartinCostello.Api.Controllers.ToolsControllerTests.Tools_Post_Hash_Returns_Correct_Response(String algorithm, String format, String plaintext, String expected)
```

**Comment by ivaylokenov** [comment]
@martincostello Well, that's interesting. I will investigate further. It is something coming from the .NET APIs as I am using their code. Thank you reporting.

**Comment by ivaylokenov** [comment]
@martincostello I will keep this open until I fix the Travis CI issue. Thank you for your feedback!

**Comment by ivaylokenov** [comment]
Hi @martincostello,

The library is now ported to .NET Core 2. I will try to explain why it was delayed by so much time.

Back in March 2017 when .NET Core migrated to the new (old) .csproj file format, I was having difficulties with the project as it can be seen here - #270 (comment)

Visual Studio was continually crashing, tests did not run correctly, and the overall developing experience was terrible. I tried for 2-3 months to port the library but could not succeed.

During that time, I started developing severe health issues from staying in front of my computer too much. I was working 12 to 16 hours from Monday to Sunday with not many breaks for 2-3 years straight. I had to stop doing it for an indefinite time. My GitHub activity shows this pretty clearly: https://github.com/ivaylokenov

The truth is - I like this testing library and have much fun developing it, so I was confused and did not know what to tell people about its future maintenance. I did not want to abandon it forever but did not know when I would be able to work actively again on it. I now realize that it was a mistake to be silent all this time and should have put an explanation earlier. The fact that my personal life was very stressful for the last two years is not an excuse.

Back in October 2018, I opened the code and started working actively on bringing back the library to life. It took me 6 months because I added lots of new features like third-party DI support, all possible action results, every single action filter, lots of helper methods for easier testing, more straightforward configuration, performance fixes and more. Version 2.2 is available on NuGet already, and version 2.1 will be published in the next few days. I will then update the tutorial, documentation, the web site, and will start working on .NET Core 3.

I am sorry for the inconvenience caused by this library and me. I truly am.

As compensation (if that is even possible), all versions before 3.0 do not have any limitations and can be used without any license code.

The good news is that I found a small private investor who is sponsoring the project. I can actively put 20 to 30 hours per week in the development of MyTested.AspNetCore.Mvc. If I manage to increase the usage and find clients, the investor will increase the budget and I will be able to hire another developer to help me. Additionally, I will put a sponsorship program with logos, and hopefully, I will be able to maintain the project steadily in the years to come.

Thank you for reading and understanding.

Ivaylo
